pkg/controller/staticpod/staticpod_controller.go

```diff
@@ -0,0 +1,662 @@
+		return r.updateStaticPodStatus(instance, totalNumber, totalNumber, totalNumber, upgradeSuccessCondition)
+	}
+
+	// The latest hash value for static pod spec
+	// This hash value is used in three places
+	// 1. Automatically added to the annotation of static pods to facilitate checking if the running static pods are up-to-date
+	// 2. Automatically added to the annotation of worker pods to facilitate checking if the worker pods are up-to-date
+	// 3. Added to static pods' corresponding configmap to facilitate checking if the configmap is up-to-date
+	latestHash := util.ComputeHash(&instance.Spec.Template)
+
+	// The latest static pod manifest generated from user-specified template
+	// The above hash value will be added to the annotation
+	latestManifest, err := util.GenStaticPodManifest(&instance.Spec.Template, latestHash)
+	if err != nil {
+		klog.Errorf(Format("Fail to generate static pod manifest of StaticPod %v, %v", request.NamespacedName.Name, err))
+		return ctrl.Result{}, err
+	}
+
+	// Sync the corresponding configmap to the latest state
+	if err := r.syncConfigMap(instance, latestHash, latestManifest); err != nil {
+		klog.Errorf(Format("Fail to sync the corresponding configmap of StaticPod %v, %v", request.NamespacedName.Name, err))
+		return ctrl.Result{}, err
+	}
+
+	// Complete upgrade info
+	{
+		// Count the number of upgraded nodes
+		upgradedNumber = setUpgradeNeededInfos(upgradeInfos, latestHash)
+
+		// Set node ready info
+		if err := checkReadyNodes(r.Client, upgradeInfos); err != nil {
+			klog.Errorf(Format("Fail to check node ready status of StaticPod %v,%v", request.NamespacedName.Name, err))
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Sync worker pods
+	var allSucceeded bool
+	deletePods := make([]*corev1.Pod, 0)
+	succeededNodes := make([]string, 0)
+	{
+		// Deal with worker pods
+		allSucceeded, err = r.checkWorkerPods(upgradeInfos, latestHash, &deletePods, &succeededNodes)
+		if err != nil {
+			klog.Errorf(Format("Fail to continue upgrade, cause worker pod of StaticPod %v in node %v failed",
+				request.NamespacedName.Name, err.Error()))
+			return r.updateStaticPodStatus(instance, totalNumber, instance.Status.DesiredNumber, upgradedNumber,
+				util.UpgradeFailedConditionWithNode(err.Error()))
+		}
+
+		// Verify succeededNodes
+		if instance.Spec.UpgradeStrategy.Type == appsv1alpha1.AutoStaticPodUpgradeStrategyType {
+			ok, err := r.verifySucceededNodes(instance, succeededNodes, latestHash)
+			if util.IsFailedError(err) {
+				klog.Errorf(Format("Fail to verify succeededNodes of StaticPod %v, %v", request.NamespacedName.Name, err))
+				return r.updateStaticPodStatus(instance, totalNumber, instance.Status.DesiredNumber, upgradedNumber,
+					util.UpgradeFailedConditionWithNode(err.Error()))
+			}
+
+			if err != nil {
+				klog.Errorf(Format("Fail to verify succeededNodes of StaticPod %v, %v", request.NamespacedName.Name, err))
+				return reconcile.Result{}, err
+			}
+
+			if !ok {
```

Comment:
I'm still working on how to control the retry times of reconcile.
The purpose of this function is to ensure that the current round of upgrade is fully completed before continuing in auto mode.

Reply:
This is out-of-date. Currently, verification will be done by upgrade worker pod.